pkg/controllers/managedzone/managedzone_controller.go

```diff
@@ -0,0 +1,371 @@
+	"sigs.k8s.io/controller-runtime/pkg/log"
+
+	v1 "github.com/Kuadrant/multi-cluster-traffic-controller/pkg/apis/v1"
+	"github.com/Kuadrant/multi-cluster-traffic-controller/pkg/dns"
+)
+
+const (
+	ManagedZoneFinalizer = "kuadrant.io/managed-zone"
+)
+
+// ManagedZoneReconciler reconciles a ManagedZone object
+type ManagedZoneReconciler struct {
+	client.Client
+	DNSProvider dns.Provider
+	Scheme      *runtime.Scheme
+}
+
+var Clock clock.Clock = clock.RealClock{}
+
+//+kubebuilder:rbac:groups=kuadrant.io,resources=managedzones,verbs=get;list;watch;create;update;patch;delete
+//+kubebuilder:rbac:groups=kuadrant.io,resources=managedzones/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=kuadrant.io,resources=managedzones/finalizers,verbs=update
+
+func (r *ManagedZoneReconciler) Reconcile(ctx context.Context, req ctrl.Request) (ctrl.Result, error) {
+	_ = log.FromContext(ctx)
+
+	previous := &v1.ManagedZone{}
+	err := r.Client.Get(ctx, client.ObjectKey{Namespace: req.Namespace, Name: req.Name}, previous)
+	if err != nil {
+		if err := client.IgnoreNotFound(err); err == nil {
+			return ctrl.Result{}, nil
+		} else {
+			return ctrl.Result{}, err
+		}
+	}
+	managedZone := previous.DeepCopy()
+
+	log.Log.V(3).Info("ManagedZoneReconciler Reconcile", "managedZone", managedZone)
+
+	if managedZone.DeletionTimestamp != nil && !managedZone.DeletionTimestamp.IsZero() {
+		if err := r.deleteParentZoneNSRecord(ctx, managedZone); err != nil {
+			log.Log.Error(err, "Failed to delete parent Zone NS Record", "managedZone", managedZone)
+			return ctrl.Result{}, err
+		}
+		if err := r.deleteManagedZone(ctx, managedZone); err != nil {
+			log.Log.Error(err, "Failed to delete ManagedZone", "managedZone", managedZone)
+			return ctrl.Result{}, err
+		}
+		controllerutil.RemoveFinalizer(managedZone, ManagedZoneFinalizer)
+
+		err = r.Update(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+		return ctrl.Result{}, nil
+	}
+
+	if !controllerutil.ContainsFinalizer(managedZone, ManagedZoneFinalizer) {
+		log.Log.Info("ManagedZoneReconciler Reconcile: Added finalizer and set parent zone owner", "managedZone", managedZone.Name)
+
+		controllerutil.AddFinalizer(managedZone, ManagedZoneFinalizer)
+
+		err = r.setParentZoneOwner(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+
+		err = r.Update(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	var reason, message string
+	status := v1.ConditionTrue
+
+	// Publish the managed zone
+	err = r.publishManagedZone(ctx, managedZone)
+	if err != nil {
+		status = v1.ConditionFalse
+		reason = "ProviderError"
+		message = fmt.Sprintf("The DNS provider failed to ensure the managed zone: %v", err)
+
+		err = r.Status().Update(ctx, managedZone)
+		if err != nil {
+			return ctrl.Result{}, err
+		}
+	}
+
+	// Create the parent zone NS record
+	err = r.createParentZoneNSRecord(ctx, managedZone)
```

Comment:
Is the usecase for the parent zone, if we want to setup a new zone under our default zone we do this by creating the new managed zone resource and pointing it at the parent and this will result in a NS record in the parent and a new zone for the child?

Reply:
Yes. This was going to be required when we were looking at a zone per tenant, and i had already written the code. We could probably get away without it for M2 now that we are just going for a single MZ tbh, but it is handy for creating zones for testing purposes.